infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java

```diff
@@ -357,50 +360,104 @@ private SequencerStatus getPrimarySequencerStatus(Layout layout, PollReport poll
-            // These conditions are mutually exclusive. If there is a failure to be
-            // handled, we don't need to explicitly fix the unfilled layout slot. Else we do.
-            Layout layout = serverContext.copyManagementLayout();
-            Set<String> failedNodes = new TreeSet<>(recommendationEngine
-                    .failedServers(clusterStateContext.getClusterState(), layout));
-
-            if (!failedNodes.isEmpty() || isCurrentLayoutSlotUnFilled(pollReport)) {
-
-                log.info("Detected failed nodes in node responsiveness: Failed:{}, pollReport:{}",
-                        failedNodes, pollReport);
-                getCorfuRuntime().getLayoutView().getRuntimeLayout(layout)
-                        .getManagementClient(serverContext.getLocalEndpoint())
-                        .handleFailure(pollReport.getPollEpoch(), failedNodes).get();
-
-            } else if (!getPrimarySequencerStatus(layout, pollReport)
-                    .equals(SequencerStatus.READY)) {
-                // If failures are not present we can check if the primary sequencer has been
-                // bootstrapped from the heartbeat responses received.
-                if (sequencerNotReadyCounter == null
-                        || sequencerNotReadyCounter.getEpoch() != layout.getEpoch()) {
-                    // If the epoch is different than the poll epoch, we reset the timeout state.
-                    sequencerNotReadyCounter = new SequencerNotReadyCounter(layout.getEpoch(), 1);
-
-                } else if (sequencerNotReadyCounter.getEpoch() == layout.getEpoch()) {
-                    // If the epoch is same as the epoch being tracked in the tuple, we need to
-                    // increment the count and attempt to bootstrap the sequencer if the count has
-                    // crossed the threshold.
-                    sequencerNotReadyCounter.setCounter(sequencerNotReadyCounter.getCounter() + 1);
-                    if (sequencerNotReadyCounter.getCounter() >= SEQUENCER_NOT_READY_THRESHOLD) {
-                        // Launch task to bootstrap the primary sequencer.
-                        log.info("Attempting to bootstrap the primary sequencer.");
-                        // We do not care about the result of the trigger.
-                        // If it fails, we detect this again and retry in the next polling cycle.
-                        getCorfuRuntime().getLayoutManagementView()
-                                .asyncSequencerBootstrap(layout, detectionTaskWorkers);
-                    }
-                }
+            ClusterState clusterState = pollReport.getClusterState();
+
+            if (clusterState.size() != layout.getAllServers().size()) {
+                log.error("Cluster representation is different than layout. Cluster: {}, layout: {}", clusterState, layout);
+                return false;
             }
 
+            Optional<NodeRank> maybeFailedNode = advisor.failedServer(
```

Comment:
My understanding is: cluster degradation should happen gradually, we found a failed node we mark it unresponsive, there could be situations when a node makes a wrong decision. If a wrong decision was made by a node it affects only another node in a cluster. 
If node accidentally can make many other nodes unresponsive it can affect cluster stability.

My approach: if 2 nodes are down they will be excluded for 2 steps it's 2 seconds, but the decision will be more accurate.

Reply:
I am not in agreement. If the cluster destabilizes because the algorithm is wrong then we fix the algorithm. Running it one step at a time is adding another fundamental new assumption to the algorithm.  Why is the decision more accurate in 2 steps vs one. Also if 2 nodes are down why is one node excluded before the second one.

Reply:
like I said before this muddles the algorithm. I do not think this improves determinism in any way. It might give the semblance of simplicity but does not make it deterministic. As regards it being more easily debuggable, I think we need to come up with better ways to debug, log actions. Also if the action engine carries out recommendations one at a time or takes into account all pending actions, we can think of doing that. But like you have used this argument a number of places, we can do that in the next iteration or version.

Reply:
ok